### DISCUSSION TITLE
Cannot convert from 'NotLoaded<T>' to 'Func<Loaded<T>>' when calling OptionT>.Match()

### DISCUSSION THREAD

**Discussion by MrYossu** [louthy/language-ext#943] [Q&A]
I'm sure I'm doing something obviously stupid, but I can't see what.

I have the following code that generates a discriminated union...

```c#
  [Union]
  public interface LoaderContainer<T> {
    LoaderContainer<T> Loading();
    LoaderContainer<T> Loaded(T value);
    LoaderContainer<T> NotLoaded();
  }
```

I want to create an extension method that will convert an `Option<T>` to ` Loaded<T>` if the `Option<T>` is `Some` and to `NotLoaded<T>` if the `Option<T>` is `None`.

Seems simple enough...

```c#
   public static LoaderContainer<T> ToLoaderContainer<T>(this Option<T> o) =>
      o.Match(val => new Loaded<T>(val), new NotLoaded<T>());
```

However, this gives a compiler error on the second parameter to `Match()` (namespaces removed to make it easier to read)...

_Cannot convert from 'NotLoaded<T>' to 'Func<Loaded<T>>'_

I tried changing the second parameter to be a `Func<T>`...

```c#
    public static LoaderContainer<T> ToLoaderContainer<T>(this Option<T> o) =>
      o.Match(val => new Loaded<T>(val), () => new NotLoaded<T>());
```

...but this gave a compiler error on the whole second line...

_Cannot implicitly convert type 'LanguageExt.Unit' to 'LoaderContainer<T>'_

What am I doing wrong? Thanks

**Comment by louthy** [comment]
I’m not in front of my computer, so not 100% sure, but it seems the compiler can’t resolve a common type, try constructing them with

    LoaderContainer.Loaded(…) and LoaderContainer.NotLoaded(…)

**Comment by louthy** [comment]
In front of my PC now, my suspicions were correct.  These are your options:

Tell C# the common join type, by providing the generic return argument to `Match`:
```c#
    public static LoaderContainer<T> ToLoaderContainer<T>(this Option<T> o) =>
        o.Match<LoaderContainer<T>>(val => new Loaded<T>(val), new NotLoaded<T>());
```
Tell C# the common join-type by casting one of the branches to the base type:
```c#
        public static LoaderContainer<T> ToLoaderContainer<T>(this Option<T> o) =>
        o.Match(val => new Loaded<T>(val) as LoaderContainer<T>, new NotLoaded<T>());
```
Use the constructor functions, rather than `new`, to create your cases.  The constructor functions return the `LoaderContainer<T>` base-type, not the more specific case-types.  This is using point-free style:
```c#
    public static LoaderContainer<T> ToLoaderContainer<T>(this Option<T> o) =>
        o.Match(LoaderContainer.Loaded<T>, LoaderContainer.NotLoaded<T>);
```
